lib/charms/operator_libs_linux/v0/sysctl.py

```diff
@@ -0,0 +1,237 @@
+        return self.args[0]
+
+class SysctlError(Error):
+    """Raised when there's an error running sysctl command."""
+
+
+class SysctlPermissionError(Error):
+    """Raised when there's an error applying values in sysctl."""
+
+
+class ValidationError(Error):
+    """Raised when there's an error in the validation process."""
+
+
+@dataclass()
+class ConfigOption:
+    """Definition of a config option.
+    
+    NOTE: this class can be extended to handle the rule tied to each option.
+    """
+    name: str
+    value: int
+
+    @property
+    def string_format(self) -> str:
+        return f"{self.name}={self.value}"
+
+
+class SysctlConfig:
+    """Represents the state of the config that a charm wants to enforce."""
+
+    def __init__(self, config_params: dict[str, dict], app_name: str) -> None:
+        self.config_params = config_params
+        self.app_name = app_name
+
+    @property
+    def config_params(self) -> list[ConfigOption]:
+        """Config options passed to the lib."""
+        return self._config_params
+
+    @config_params.setter
+    def config_params(self, value: dict):
+        result = []
+        for k, v in value.items():
+            result += [ConfigOption(k, v["value"])]
+        self._config_params = result
+
+    @property
+    def charm_filepath(self) -> str:
+        """Name for resulting charm config file."""
+        return f"{SYSCTL_DIRECTORY}/90-juju-{self.app_name}"
+
+    @property
+    def charm_config_exists(self) -> bool:
+        """Return whether the charm config file exists."""
+        return os.path.exists(self.charm_filepath)
+
+    @property
+    def merged_config_exists(self) -> bool:
+        """Return whether a merged config file exists."""
+        return os.path.exists(SYSCTL_FILENAME)
+
+    @property
+    def merged_config(self) -> list[ConfigOption] | None:
+        """Return applied internal config."""
+        if not self.merged_config_exists:
+            return None
+
+        with open(SYSCTL_FILENAME, "r") as f:
+            return [ConfigOption(param.strip(), int(value.strip()))
+                    for line in f.read().splitlines()
+                    if line and not line.startswith('#')
+                    for param, value in [line.split('=')]]
+
+    def update(self) -> None:
+        """Update sysctl config options."""
+        if self.charm_config_exists:
+            return
+        self._create_charm_file()
+
+        if not self.validate():
+            raise ValidationError()
+
+        snapshot = self._create_snapshot()
+        logger.debug(f"Created snapshot for keys: {snapshot}")
+        try:
+            self._apply()
+        except SysctlPermissionError:
+            self._restore_snapshot(snapshot)
+            raise
+        except ValidationError:
+            raise
+        self._merge()
+
+    def validate(self) -> bool:
+        """Validate the desired config params against merged ones."""
+        return True
```

Comment:
Here you want to check if any values changes? e.g. `if current_value != new_value: raise ValidationError`